### ISSUE TITLE
Linux CI for homebrew-core

### ISSUE THREAD

**Issue by iMichka** [Homebrew/brew#10597]
**Provide a detailed description of the proposed feature**

Current state:
- We have linux CI in linuxbrew-core, using github actions + Docker
- We have linux CI in homebrew-core for linux-only formulae, using github actions + Docker 

Until now, we were able to build all packages, even the heaviest ones (like llvm or rust), and the compilation speed if fine.
Basically, Github actions are easy to use, are fast, are secure, and we do not need to do any admin work.

I propose to add self-hosted runners for Linux. We need at least as many as the Mac Intel workers, but more would be better.

Requirements:

- 1 job needs at least 4 cores, 8 Gb Ram, 50 to 60 Gb disk space, and a timeout of 72 hours.
- We would like to ask for sponsorship to avoid having to pay for the setup

Bonus:

If possible, the setup of the self-hosted runners should be as easy as possible.
From my experience, I know that navigating through cloud platforms can be tedious.

If some cloud provider has an on-the-shelf 3 click option to make the setup, this is what we need.
I would like to avoid vagrant, helm, terraform, and all that stuff, which is complex (if you do not have hours to spend on the topic).

Bonus2 (Security):

If we go for VMs, the github action will need to trigger a VM deployment on the fly, and tear it down at the end.
We need to be extra careful when handling secrets, especially as we will run this from pull requests. This might be an issue.

Possible cloud providers:

- Digital Ocean
- Amazon
- Microsoft
- Google
- Others?

I would like to gather some thoughts before starting to contact cloud providers for sponsorhip.

I would like to move forward on that topic: I was hoping to find a quick solution months ago but it did not play out. We can't leave homebrew-core without linux CI much longer as the diff between both repos is now quickly getting smaller.

**What is the motivation for the feature?**

This is the last blocking part for the linuxbrew-core to the homebrew-core migration.

The timeout on Github actions is 6 hours. As soon as we will be starting to run more builds in homebrew-core, we will hit that limit pretty fast.

This was previously not an issue in linuxbrew-core as each bottle is built separately in a single job.

**How will the feature be relevant to at least 90% of Homebrew users?**

It's part of the linuxbrew-core to homebrew-core migration project.

**What alternatives to the feature have been considered?**

- Wait for github to lift the 6 hours limit. This has only advantages, as it means we have nothing to change in our setup. We have contacted github regarding this but have no visibility if this is possible: we might get an answer in a few weeks, or in 6 months, or never.
- Completely move away from github actions (that would be sad). We were using Azure Pipelines at one point, so it's doable. But Azure Pipelines have limitations too (for the free tier).

Finances:
- Another option is to hire someone to do the setup if we do not want to spend time with this
- We can pay for the hardware

**Comment by sjackman** [comment]
> I propose to add self-hosted runners for Linux. We need at least as many as the Mac Intel workers, but more would be better.

Homebrew for macOS builds bottles for four separate targets, whereas Homebrew for Linux builds for one sole target. Should one quarter of the workers suffice?

**Comment by sjackman** [comment]
> The timeout on Github actions is 6 hours. As soon as we will be starting to run more builds in homebrew-core, we will hit that limit pretty fast.

Most (>99%?) of formulae build within 6 hours. Can we use a hybrid model where typical formulae use GitHub Actions hosted runners, and a list of naughty formulae use self-hosted runners?

**Comment by sjackman** [comment]
> Possible cloud providers: Digital Ocean

https://www.digitalocean.com/open-source/

We have an existing relationship with DigitalOcean, who previously sponsored our Linux Jenkins workers. That seems like a good place to start.

**Comment by sjackman** [comment]
Thank you for starting this discussion, Michka! I am *very* excited for the merging of the cores. 🍏 💞 🐧

**Comment by dawidd6** [comment]
Timeout related stuff is mostly considered because of reverse dependencies testing, so jobs like: Python, Go, Rust and others. They take a long time to finish, but as of now it wasn't a problem in linuxbrew-core, because we didn't test reverse dependencies. This is going to change for Linux in homebrew-core. I think it is good to explain this here too.

I have no idea how to setup runners that would be spun up and down on each new workflow run nor I haven't seen any official solution from GitHub to achieve that. One thing I saw is `--once` option that could be passed to `run.sh` (runner's run script), which means the runner will exit after one workflow job. This could be of use for us. I'm thinking we could make a simple script that would execute those steps in a loop: create Docker container, run the runner in it, remove the container. Last time I checked, the runner executes just fine in our homebrew/brew container. This way we don't need to provision new VMs each workflow run, but just containers. Don't know if this is really what we would want, considering security and all those quirks, but I think it would be a decent setup.

**Comment by issyl0** [comment]
> If some cloud provider has an on-the-shelf 3 click option to make the setup, this is what we need. I would like to avoid vagrant, helm, terraform, and all that stuff, which is complex (if you do not have hours to spend on the topic).

Whatever we have needs to be reproducible though, be it an instance spun up with Terraform, and/or cloud-init/some other configuration management software to configure what's on it. As "easy" as it may seem, pointing and clicking is not really that - cloud provider UIs have a dizzying array of possible options. With the "configuration as code" way, we *know* that we're doing things the same every time we have to re-make one of these instances - which is important for security as well as maintainers' confidence levels in the system's ability to meet our needs.

I am, however, well aware that I am saying this as someone who has been working with these technologies for years. I am also well aware that we're all volunteers...

> Another option is to hire someone to do the setup if we do not want to spend time with this

GSoC? Or an actual contractor? If the former: I'm willing to mentor that. If the latter: I know lots of good people. 😉

**Comment by MikeMcQuaid** [comment]
> Homebrew for macOS builds bottles for four separate targets, whereas Homebrew for Linux builds for one sole target. Should one quarter of the workers suffice?

Agreed.

> If we go for VMs, the github action will need to trigger a VM deployment on the fly, and tear it down at the end.
> We need to be extra careful when handling secrets, especially as we will run this from pull requests. This might be an issue.

This shouldn't be an issue, we just can't have it be in the same `workflow` file but instead one based off `pull_request_target` or similar.

> Whatever we have needs to be reproducible though, be it an instance spun up with Terraform, and/or cloud-init/some other configuration management software to configure what's on it. As "easy" as it may seem, pointing and clicking is not really that - cloud provider UIs have a dizzying array of possible options. With the "configuration as code" way, we _know_ that we're doing things the same every time we have to re-make one of these instances - which is important for security as well as maintainers' confidence levels in the system's ability to meet our needs.

I agree. I specifically want to avoid having anything non-automatable. Our macOS setup should be more automated than it is but the state of automation (both in terms of cloud providers and being able to do e.g. all updates over SSH) is much more limited in macOS than Linux.

> The timeout on Github actions is 6 hours. As soon as we will be starting to run more builds in homebrew-core, we will hit that limit pretty fast.

An interim option we could consider is simply not testing reverse dependencies on Linux at all (or triggering one or more extra jobs to do so which would also be useful on macOS).

> * Completely move away from github actions (that would be sad). We were using Azure Pipelines at one point, so it's doable. But Azure Pipelines have limitations too (for the free tier).

Hard 👎🏻 to moving away from GitHub Actions for me.

> Most (>99%?) of formulae build within 6 hours. Can we use a hybrid model where typical formulae use GitHub Actions hosted runners, and a list of naughty formulae use self-hosted runners?

Yes, we should do this. This would also be useful on macOS for e.g. everything that's `bottle :unneeded` and we feel like we don't need to test it on multiple platforms.

> I have no idea how to setup runners that would be spun up and down on each new workflow run nor I haven't seen any official solution from GitHub to achieve that. One thing I saw is `--once` option (which isn't mentioned officially) that could be passed to `run.sh` (runner's run script), which means the runner will exit after one workflow job. This could be of use for us. I'm thinking we could make a simple script that would execute those steps in a loop: create Docker container, run the runner in it, remove the container. Last time I checked, the runner executes just fine in our homebrew/brew container. This way we don't need to provision new VMs each workflow run, but just containers. Also with that solution we won't have to duplicate the workflow files solely because of Linux jobs needing a container. Don't know if this is really what we would want, considering security and all those quirks, but I think it would be a decent setup.

My thinking on both macOS and Linux is that we need to set up some sort of system that means a VM can auto connect itself as a self-hosted runner on boot. This may require speaking to a simple intermediate service that can decide e.g. what number/name/token this self-hosted runner requires. CC @jonchang who has had some thoughts/work on this in the past.

---

Just to step back: I think a lot of this work is worth thinking about whether it would improve things on macOS, too. There's room to improve the CI and operations experience pretty dramatically on both platforms through an effort like this.

**Comment by Bo98** [comment]
> An interim option we could consider is simply not testing reverse dependencies on Linux at all

If we absolutely need to then sure, but the goal should still be to test reverse dependencies on Linux as well due to different dependency graphs as well as differences in how things link (e.g. Python).

While testing reverse dependencies is a time-consuming job (well over 24 hours for Rust), it is also incredibly invaluable in making sure nothing breaks (Rust updates usually break something).

**Comment by iMichka** [comment]
> Homebrew for macOS builds bottles for four separate targets, whereas Homebrew for Linux builds for one sole target. Should one quarter of the workers suffice?

Err yes I was maybe unclear: if there are 4 workers for Big Sur, 4 for Catalina, and so on, then we should just need 4 for Linux. I think it's ok to have the same number of workers for each platform. But we might need less, see below:

> Most (>99%?) of formulae build within 6 hours. Can we use a hybrid model where typical formulae use GitHub Actions hosted runners, and a list of naughty formulae use self-hosted runners?

Yes, my experiment with self hosted runners in linuxbrew-core shoved that you can choose the worker you want based on a github label. Given this, we might need less Linux workers than mac workers. The label can be added manually, or based on a curated list of formulae that we know will take more than 6 hours (there are not that many).

>  create Docker container, run the runner in it, remove the container. Last time I checked, the runner executes just fine in our homebrew/brew container. This way we don't need to provision new VMs each workflow run, but just containers.

Yes that's also an alternative: we can use an always-on VM, and instead of spining it up and down, we can spawn containers on it. Of course you'll pay for VM the whole time, even when it's idle. On the other hand I have the feeling that our runners are never really idle in homebrew-core.

> Whatever we have needs to be reproducible though, be it an instance spun up with Terraform, and/or cloud-init/some other configuration management software to configure what's on it. As "easy" as it may seem, pointing and clicking is not really that - cloud provider UIs have a dizzying array of possible options.

What I was thinking is that maybe there is a company out there that does exctaly what we need, but as a service. Basically provisioning self-hosted runners for builds, without too much hassle. I should start my own startup maybe? But github could just kill my startup by lifting the timeout time or providing a few runner options in their settings, in exchange of a few dollars.

I am willing to learn a little bit of terraform, I was just hoping maybe someone out there would provide an easier solution.

> GSoC? Or an actual contractor? If the former: I'm willing to mentor that. If the latter: I know lots of good people.

GSoc is probably too late IMHO, I would like to start this project in the next weeks. For the recruitment, I want to explore first how difficult it is, and go down that route only if required, because it involves another level of paperwork I have no vision on.

> An interim option we could consider is simply not testing reverse dependencies on Linux at all 

As some others said above, I think we really do want to test reverse dependencies, we really want to avoid breakages.

> a VM can auto connect itself as a self-hosted runner on boot.

That's what I did with my test VM on azure: it would spin up and launch the github deamon, and register itself. That was for one VM, but you can do the same for multiple VMs.

**Comment by MikeMcQuaid** [comment]
> Yes that's also an alternative: we can use an always-on VM, and instead of spining it up and down, we can spawn containers on it. Of course you'll pay for VM the whole time, even when it's idle. On the other hand I have the feeling that our runners are never really idle in homebrew-core.

They are definitely idle at times. Also, if we're able to spin them up and down we can do more parallelisation which will exacerbate this.

> What I was thinking is that maybe there is a company out there that does exctaly what we need, but as a service. Basically provisioning self-hosted runners for builds, without too much hassle. I should start my own startup maybe? But github could just kill my startup by lifting the timeout time or providing a few runner options in their settings, in exchange of a few dollars.

Yeh, I think pretty much all the major cloud platforms provide some way of doing "run something in a custom docker image without caring about the machine underneath".

> For the recruitment, I want to explore first how difficult it is, and go down that route only if required, because it involves another level of paperwork I have no vision on.

I can help with the writeup for paperwork.



> As some others said above, I think we really do want to test reverse dependencies, we really want to avoid breakages.
> That's what I did with my test VM on azure: it would spin up and launch the github deamon, and register itself. That was for one VM, but you can do the same for multiple VMs.

👍🏻

**Comment by iMichka** [comment]
I made a small attempt using google cloud (because my azure free credits ran out so I can't test azure anymore).
From an usability perspective, I found google cloud easier to use compared to azure.

This blog post helped a lot (though I used a compute engine (VM) instead of an app engine):
https://github.blog/2020-08-04-github-actions-self-hosted-runners-on-google-cloud/

I have now a local .sh script that setups a gcloud VM. The VM will have docker installed and runs the GitHub Actions runner.
So far so good.

I learned two things:

1) I learned that you can get a new self-hosted GitHub token (let's call it VM_TOKEN) using the GitHub API (https://api.github.com/repos/${OWNER}/${REPO}/actions/runners/registration-token). This is the VM token you would normally get when you add a new runner in the GitHub Settings. This is nice, as it avoids to need to got the Github GUI and add a new runner manually. This operations needs a HOMEBREW_GITHUB_API_TOKEN though, so I did this on my local script (to avoid to have to deploy HOMEBREW_GITHUB_API_TOKEN on the VM). The VM_TOKEN is deployed to the self-hosted runner though (which means that anyone opening a Pull Request can get the VM token, which allows him to unregister the VM. I think the token has a lifetime of 60 minutes for the addition of a new VM with the same token, so I think nobody can impersonate our VM).

The drawback is that if you want to delete the VM completely (for example when you want to renew it), you need this VM_TOKEN, and you need to run `./config.sh remove --token VM_TOKEN` from within the actions runner directory in the VM. So the best is to 'force remove' the runner manually from the GitHub UI.

It's not a big deal but redeploying a new VM involves a manual step, so it is not 100% automatised.

2. To start/stop the VM from the CI, I planned to do the following: run a first step on `ubuntu-latest` that starts the VM, run the second step on `self-hosted` (using a docker container on the VM), and a third step on `ubuntu-latest` to stop the VM.
For the start step, in can use `gcloud compute instances list` to find a stopped VM, and then I can use `gcloud compute instances start VM_NAME` to start the VM. Then I can probably have a script that waits for the VM to boot, but it's pretty fast.

One issue that might arise is race conditions when VM's are started, if multiple jobs try to start the same VM. Not sure this is an actual issue.

Another issue is that these gcloud operations involve google credentials. I found https://github.com/google-github-actions/setup-gcloud to setup the credentials, but this involves exposing our google cloud credentials in the workflow, so anyone can open a pull request and get our credentials and compromise our cloud infrastructure.

Thoughts?

**Comment by MikeMcQuaid** [comment]
> Another issue is that these gcloud operations involve google credentials. I found https://github.com/google-github-actions/setup-gcloud to setup the credentials, but this involves exposing our google cloud credentials in the workflow, so anyone can open a pull request and get our credentials and compromise our cloud infrastructure.

Can use `pull_request_target` to do anything sensitive using the `master` branch so users cannot compromise our workflows/infrastructure/tokens. This can then either kick off a `pull_request` job or do the checkout manually at a later stage in `pull_request_target` when we've destroyed/aren't exposing the necessary secrets.

The only downside of this is this job would need to also run on non-master `push` events to maintainers can test workflow changes as otherwise a change to a workflow would only take effect after merge rather than before.



> 1. I learned that you can get a new self-hosted GitHub token (let's call it VM_TOKEN) using the GitHub API ([https://api.github.com/repos/${OWNER}/${REPO}/actions/runners/registration-token](https://api.github.com/repos/$%7BOWNER%7D/$%7BREPO%7D/actions/runners/registration-token)). This is the VM token you would normally get when you add a new runner in the GitHub Settings. This is nice, as it avoids to need to got the Github GUI and add a new runner manually. This operations needs a HOMEBREW_GITHUB_API_TOKEN though, so I did this on my local script (to avoid to have to deploy HOMEBREW_GITHUB_API_TOKEN on the VM). The VM_TOKEN is deployed to the self-hosted runner though (which means that anyone opening a Pull Request can get the VM token, which allows him to unregister the VM. I think the token has a lifetime of 60 minutes for the addition of a new VM with the same token, so I think nobody can impersonate our VM).

This is very nice. @jonchang @Bo98 any ideas if this is something we could start doing with MacStadium too?

**Comment by Bo98** [comment]
> So the best is to 'force remove' the runner manually from the GitHub UI.

If it helps, you can replace runners on startup with `--replace`.

> any ideas if this is something we could start doing with MacStadium too?

We talked a bit about this last year. At the time, it wasn't easy to pass in secrets to Orka but it was something they were looking to implement as a feature. The only real option was extra infrastructure to deal with the secrets.

This did predate the existence of `pull_request_target` though. How does one run a `pull_request_target` initially and then de-escalate to a `pull_request` workflow? I know the opposite is possible with `workflow_run`.

**Comment by MikeMcQuaid** [comment]
> We talked a bit about this last year. At the time, it wasn't easy to store and pass in secrets to Orka but it was something they were looking to implement as a feature. The only real option was extra infrastructure to deal with the secrets (since we couldn't have them in our workflow at all).

As long as you don't need a worker-specific secret I think this is something we could do already. My thinking (and please correct this) is:
- machine boots from an `orka vm deploy`
- machine uses GitHub secrets as `brewadmin` user to register itself as a GitHub Actions worker on boot/resume (with a unique ID)
- machine runs jobs as `brew` user with no access to GitHub secrets
- machine can be killed by GitHub Actions querying the API, seeing it's inactive and running the specific `orka` commands to destroy and rebuild it. This same job can remove workers from GitHub that are inactive because they've already been destroyed or those that have crashed and are unresponsive by running `orka` commands.

This could initially be a manually run GitHub Actions job to run all the relevant `orka` commands without secrets access, become a (increasingly regular) scheduled job and eventually (if we can get it fast enough) be something we can do on every build.

Thoughts? Concerns? What pieces are missing here?

**Comment by MikeMcQuaid** [comment]
> This did predate the existence of `pull_request_target` though. How does one run a `pull_request_target` initially and then de-escalate to a `pull_request` workflow? I know the opposite is possible with `workflow_run`.

You can do so manually be checking out the actual pull request within the workflow.

**Comment by Bo98** [comment]
> machine uses GitHub secrets as `brewadmin` user to register itself as a GitHub Actions worker on boot/resume (with a unique ID)

Presumably this secret is baked into the image in a file/keychain only readable by `brewadmin`?

> You can do so manually be checking out the actual pull request within the workflow.

That won't suddenly hide all secrets though. I'm talking about de-escalating from an environment where there is write tokens to an environment where there is none. We should be avoiding doing any checking out in `pull_request_target`.

**Comment by MikeMcQuaid** [comment]
> Presumably this secret is baked into the image in a file/keychain only readable by `brewadmin`?

@Bo98 yes, that's what I was thinking.

> That won't suddenly hide all secrets though. I'm talking about de-escalating from an environment where there is write tokens to an environment where there is none. We should be avoiding doing any checking out in `pull_request_target`.

Secrets are exposed to the environment per-step in GitHub Actions. As a result, if you control the workflow file you can ensure that a given step where you checkout and run untrusted code does not have any access to secrets.

**Comment by Bo98** [comment]
> Secrets are exposed to the environment per-step in GitHub Actions.

And absolutely no tokens will be in memory somewhere?

Potential (although extreme) attack vector I was thinking was:

* Build a native program that reads memory of parent processes (the GitHub worker process)
* Commit program binary to repo
* Create a formula which `system` calls this binary.

**Comment by MikeMcQuaid** [comment]
> And absolutely no tokens will be in memory somewhere?

Correct. The environment is exposed per step (or job) per process.

**Comment by iMichka** [comment]
Thanks @MikeMcQuaid, the hint about the `pull_request_target` trigger really helped.

I have now a first workflow that starts a VM, using the `pull_request_target` trigger:
https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/start-vm.yml

The test workflow waits for a self-hosted runner to be started:
https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/tests-self-hosted.yml

I thought I would have more work to do here but that workflow just waits for a VM to spin up and register, so that is pretty simple.

The workflow then uses docker to run the build on the VM. I need to finish to configure docker on the VM but I already did it back then when testing the same on Azure, so nothing fancy here.

The last workflow stops the VM:
https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/stop-vm.yml

It uses the `workflow_run` trigger. It will wait for the previous test workflow to be completed.
It's a little bit a shame that this workflow is not displayed in the Github PR UI, because it runs on master.

TODO:
- the start workflow needs to select a stopped VM from the VM list
- I need to pass the VM's name through the 3 workflows (so that the right VM is used end to end). Don't know how hard that will be.
- I need to finish to configure the VM's: docker is not correctly setup and I think I will need to add extra hard drives.

**Comment by iMichka** [comment]
> I need to pass the VM's name through the 3 workflows (so that the right VM is used end to end). Don't know how hard that will be.

Small update: my start VM script adds a label to the pull request (for example `linux-self-hosted-1`). The self hosted workflow will fetch that label in a first job, then sets up a dynamic matrix (usin https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#fromjson), and the self hosted runner is now selected based on the label that was set by the startVM workflow.

The issue is now to stop the VM based on that label (or any other way). As my stop VM workflow uses the `workflow_run` trigger, I do not have access to the related pull request in the github context. I only have acces to the parent workflow id that triggered the `workflow_run`. Any idea how I can pass a parameter from my `pull_request` workflow to my `workflow_run` workflow?

**Comment by iMichka** [comment]
I found a way: I used `actions/upload-artifact@v2` to store the pull request number and the runner name in the temporary file storage of the workflow (using a file), and then used `dawidd6/action-download-artifact@v2` in my stop vm workflow to retrieve the files, and read the runner name and pull request number. I can stop the VM and finally cleanup the label from the pull request.

Remaining TODO:
* the start workflow needs to select a stopped VM from the VM list
* I need to finish to configure the VM's: docker is not correctly setup and I think I will need to add extra hard drives.

**Comment by iMichka** [comment]
All done, the docker set-up is now done. And the start VM script waits for a free VM to be started.

I will now make a final check on a bigger PR to see if the disk space / memory / compilation time is fine. Based on that I'll contact google to see if they can give us some credits for VMs.

On final thing, my implementation has a minor drawback: the test workflow waits for a PR to be labeled  with the runner name. This means that as soon as someone adds a label to a PR, that check workflow will be started. And once that workflow is done, the stop VM workflows is triggered. This means that each time you label a PR, 2 workflows will be triggered ... no really nice, but I found no other way.

Here are the three workflows, I would like to get some feedback about the overall strategy:

https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/start-vm.yml
https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/tests-self-hosted.yml
https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/stop-vm.yml

**Comment by MikeMcQuaid** [comment]
> Here are the three workflows, I would like to get some feedback about the overall strategy:
> 
> https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/start-vm.yml
> https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/tests-self-hosted.yml
> https://github.com/Homebrew/linuxbrew-core/blob/master/.github/workflows/stop-vm.yml

Looks pretty elegant to me! Question: why do you want to require the specific self-hosted running name to be specified rather than e.g. just selecting any available in the pool?

> This means that each time you label a PR, 2 workflows will be triggered ... no really nice, but I found no other way.

Could they be in the same workflow? Can the workflow only be triggered somehow if the relevant label is added?

**Comment by iMichka** [comment]
> Question: why do you want to require the specific self-hosted running name to be specified rather than e.g. just selecting any available in the pool?

I tried that, and it works. You can start a VM, and just let the test job pick it up from the pool when it is connected. The issue is that you may pick up a VM that just finished another job, and has not been stopped yet. So the stop job from that other job might kill the VM you just picked up. And it also means that you picked up the wrong VM, and the VM you started will not be stopped as it was not picked up. I wanted to avoid that kind of race condition by having the following rule: 1 PR = 1 specific runner

> Could they be in the same workflow?

The self-hosted test workflow is triggered by `on pull_request`, so it has not credentials because it can be run from forks. So you need a separate workflow to stop the VM.

> Can the workflow only be triggered somehow if the relevant label is added?

No. You need to spin up the stop workflow and check for the label (with something like contains(github.event.pull_request.labels.*.name, 'self-hosted-ci'). I did not implement that check in my stop workflow so I'll still have to do that.

Another solution is to use a cron job that stops idle VMs (I need to check if there is an API for that).

**Comment by iMichka** [comment]
Some options I have explored regarding sponsorship (to keep you updated and to keep track of what has been done)

# Github

I am waiting for feedback to see if the timeout could still be changed for us, right no we have no visibility on that side.

# Google

- Their sales team could not help me (but I probably talked to the wrong people)
- https://opensource.google/docs/growing/funding/ gives no instructions on how to proceed
- They did not answer to my ping on twitter

# Microsoft
 - There seems to be a funding opportunity here, but only in case some Microsoft employees are using brew on Linux, and would like to sponsor us: https://github.com/microsoft-sponsorships/microsoft-foss-fund/blob/main/README.md . I did not apply, I consider that last resort and is also just a temporary solution.
 - There seems to be something called azure sponsorship, but not sure how to get invited and if there is any long-term commitment from Microsoft. It looks more like some starting offer for startups: https://azure.microsoft.com/en-us/offers/ms-azr-0036p/

# AWS: not yet explored

**Comment by iMichka** [comment]
My final thoughts (after some discussions yesterday at the AGM):

- Finding sponsorship is not that easy
- As AWS and Azure are not so easy to use, I would recommend using Google cloud, as it was super easy to setup for our use case.
- Starting and stopping VMs in a workflow is doable, but involves 3 complex workflows and I think the solution is not that elegant (and not so readable for our users). Also, there is not so much downtime in homebrew-core, the VMs will be running almost full-time anyway.

My final proposal is to use the money from our donators to pay for 4 Google Cloud VMs that will run full-time, and we let GitHub Actions handle the dispatching of the jobs to the runners.

We will probably be able to get some statistics of the usage of the machines after a few months, and if we see that the idling time is high, we can still work on a better solution.

I'll let the newly elected @Homebrew/plc discuss this topic and come back to me with their decision.

**Comment by sjackman** [comment]
Are these self-hosted runners for the long-running jobs, or all jobs?

**Comment by iMichka** [comment]
All jobs. It would match exactly the way it is done on macOS.

If you want to reduce the number of VMs, we can have a part run on github actions; and the other one on self-hosted runners. But this means keeping a double list of formulae up to date (like we do in https://github.com/Homebrew/homebrew-core/blob/master/.github/workflows/tests-linux.yml#L2-L28). We need to include formulae in one job and exclude them in the other job.

I think we might be able to reduce the number of runners with that. But it adds maintenance burden.

**Comment by MikeMcQuaid** [comment]
> If you want to reduce the number of VMs, we can have a part run on github actions; and the other one on self-hosted runners. But this means keeping a double list of formulae up to date (like we do in https://github.com/Homebrew/homebrew-core/blob/master/.github/workflows/tests-linux.yml#L2-L28). We need to include formulae in one job and exclude them in the other job.
> 
> I think we might be able to reduce the number of runners with that. But it adds maintenance burden.

I think it'd be good to reduce the number of self-hosted runners we need. GitHub's infrastructure is preferable when possible. We could also use similar logic on macOS. Additionally, these lists will eventually go away.

I think the maintenance overhead is worth it for saving actual money here.

> * Starting and stopping VMs in a workflow is doable, but involves 3 complex workflows and I think the solution is not that elegant (and not so readable for our users). Also, there is not so much downtime in homebrew-core, the VMs will be running almost full-time anyway.

Given we have an automated way of doing this: I think having these VMs at least torn down/up daily on a scheduled job would be a nice middle ground.

**Comment by iMichka** [comment]
> I think it'd be good to reduce the number of self-hosted runners we need. GitHub's infrastructure is preferable when possible. We could also use similar logic on macOS. Additionally, these lists will eventually go away.
> I think the maintenance overhead is worth it for saving actual money here.

Ok then I'll add a few formulae to the list when I setup the workflow in homebrew-core. Maintainers may need to adapt the list from time to time.

1 runner is probably enough to start with.

> Given we have an automated way of doing this: I think having these VMs at least torn down/up daily on a scheduled job would be a nice middle ground.

Are you talking about starting / stopping the VMs or tearing them down completely? I only explored the first option. Tearing down and setting up is another kind of work. Doing it on schedule means we need an API that can check for idle VM's, and I do not know if something like this exisits. We would need to ssh into the machine and check the state of the github actions runner during the night.

**Comment by MikeMcQuaid** [comment]
> Are you talking about starting / stopping the VMs or tearing them down completely? I only explored the first option. Tearing down and setting up is another kind of work. Doing it on schedule means we need an API that can check for idle VM's, and I do not know if something like this exisits. We would need to ssh into the machine and check the state of the github actions runner during the night.

Tearing them down ideally. It would be sufficient to e.g. manually create a VM snapshot (or whatever the Google Cloud equivalent is) and restore from that.

> Doing it on schedule means we need an API that can check for idle VM's, and I do not know if something like this exisits.

I think the self-hosted runner APIs should be able to do something like this: https://docs.github.com/en/rest/reference/actions#self-hosted-runners

**Comment by iMichka** [comment]
Right, looks we can do this. Someone from the PLC (or you) need to setup a google account , so I can start setting up one single linux VM, and write the documentation for the ops.

**Comment by iMichka** [comment]
Pinging the @Homebrew/plc again: I would like the PLC to give their green light on setting up a google cloud account, so that I can add a VM for Linux self-hosted runners. Can you let me know when this will be discussed and if this is OK for you?

**Comment by issyl0** [comment]
@iMichka Sorry for the delay. I am new to PLC duties, but I've posted that question for 👍🏻 / 👎🏻 / 💬 in the PLC Slack channel and will get back to you this week!

**Comment by issyl0** [comment]
@iMichka The PLC have voted in favour of setting this up. :tada: Looking forward to seeing this in action!

**Comment by iMichka** [comment]
Thanks! I'll contact you and the PLC on Slack to discuss the setup.

**Comment by iMichka** [comment]
The final decision was taken to go for google Cloud; and use our sponsor's money to pay for it. I am currently setting up the VM and writing the documentation. Closing here as the setup is almost done.

We will run most of the builds in GitHub Actions, and have a list of formulae that need long build times that will run on the self-hosted runner. The general concept is explained here: https://github.com/Homebrew/homebrew-core/pull/72732 (this PR will be revived soon).
